### PR TITLE
Expose upsert/inject information for SQL tables

### PR DESCRIPTION
**Pull request by gierlachg**
To enable `INSERT`s for auto-discovered maps on Jet side.

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1948/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1948/Hazelcast-pr-builder-1948.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 12 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PojoUpsertTargetDescriptor.java:62:46: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PojoUpsertTargetDescriptor.java:63:50: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PojoUpsertTargetDescriptor.java:64:58: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PojoUpsertTargetDescriptor.java:77:58: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptor.java:69:50: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptor.java:70:42: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptor.java:71:40: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptor.java:72:50: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptor.java:85:44: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptor.java:86:41: '&&' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/test/java/com/hazelcast/sql/impl/inject/PojoUpsertTargetDescriptorTest.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/test/java/com/hazelcast/sql/impl/inject/PortableUpsertTargetDescriptorTest.java:1: File does not end with a newline. [NewlineAtEndOfFile]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1949/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1949/Hazelcast-pr-builder-1949.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:124 There are 3 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3895, Failures: 1, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Review by viliam-durina — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java [hunk 1]
```diff
@@ -37,6 +37,8 @@
     private final String mapName;
     private final QueryTargetDescriptor keyDescriptor;
     private final QueryTargetDescriptor valueDescriptor;
+    private final Object keyAppendix;
+    private final Object valueAppendix;
```
Comment (viliam-durina) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/AbstractMapTable.java:41]: I'd prefer something more pointing to the use, such as `keyResolverPluginData`

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import com.hazelcast.replicatedmap.ReplicatedMap;
+
+/**
+ * Plugin mechanism, allowing to perform additional analysis of
+ * {@link IMap}/{@link ReplicatedMap} keys & values on sample based schema resolution.
+ * <p/>
+ * Used by Jet.
+ */
+@FunctionalInterface
+public interface MapResolverPlugin {
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java:29]: Minor comment - the name is not very specific. I would name it something like `JetMapMetadataResolver` or alike, to highlight that is:
1. Comes from Jet
1. Deals with metadata

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Return an appendix that is attached to the resolved {@link AbstractMapTable}.
+     */
+    Object resolve(Object object, boolean key);
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java:34]: Also minor comment.

In the resolver's code, you either pass here Class of ClassDefinition as the first argument. It is ok from IMDG perspective, but are you ok with a necessity of casts inside Jet implementation? 

If you leave the interface as is, I would propose adding JavaDoc about expected arguments.

**Review by devozerov — APPROVED** [review]
Approving with two minor comments about the naming of the new interface and its contract.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java (hunk 3 above)
Comment (gierlachg) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java:34]: I had an impression that it was the proposed solution. Anyway, what about this one then?
```
interface JetMapMetadataResolver {
    Object resolveClass(Class<?> clazz, boolean key);

    Object resolvePortable(ClassDeficition clazz, boolean key);
}
```
Additional methods could be added if needed (i.e. Json).

Comment (devozerov) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/MapResolverPlugin.java:34]: They both OK from the IMDG perspective. The question is - what is better for Jet. 
The interface you proposed in the comment looks good to me - it is clean and simple.

**Review by devozerov — APPROVED** [review]

**Review by viliam-durina — APPROVED** [review]

**Pull request merged**